packages/xo-server/src/xapi/mixins/vm.mjs

```
@@ -140,53 +140,100 @@ const methods = {
+            _vdisToCreate.push(vdi)
+          }
+
+          return
+        }
+
+        vdi.$ref = this.getObject(vbd.VDI).$ref
+        if (vdi.destroy) {
+          _vdisToDestroy.push(vdi)
+          return
+        }
+
+        _vdisToUpdate.push(vdi)
+      })
+
+      // TODO: set vm.suspend_SR
+      await Promise.all([
+        ..._vdisToDestroy.map(vdi => this.VDI_destroy(vdi.$ref)),
+        // Creates the user defined VDIs.
+        ..._vdisToCreate.map((vdi, i) =>
           this.VDI_create({
-            name_description: vdiDescription.name_description,
-            name_label: vdiDescription.name_label,
-            virtual_size: vdiDescription.size,
-            SR: this.getObject(vdiDescription.sr || vdiDescription.SR, 'SR').$ref,
+            name_description: vdi.name_description,
+            name_label: vdi.name_label,
+            virtual_size: vdi.size,
+            SR: this.getObject(vdi.sr, 'SR').$ref,
           }).then(vdiRef =>
             this.VBD_create({
               // Either the CD or the 1st disk is bootable (only useful for PV VMs)
               bootable: !(hasBootableDisk || i),
 
-              userdevice: devices[i],
+              userdevice: vdi.userdevice,
               VDI: vdiRef,
               VM: vm.$ref,
             })
           )
-        )
-      )
+        ),
+        // Modify existing (previous template) disks if necessary
+        ..._vdisToUpdate.map(async ({ $ref, sr, size, ...properties }) => {
+          delete properties.destroy
+          delete properties.userdevice
+
+          await this._setObjectProperties({ $ref, $type: 'VDI' }, properties)
+
+          let _vdi = this.getObject($ref)
+          // if another SR is set, move it there
+          if (sr !== undefined) {
+            _vdi = await this.moveVdi(_vdi.$id, sr)
+          }
+
+          // if the disk is bigger
+          if (size != null && size > _vdi.virtual_size) {
+            await this.resizeVdi(_vdi.$id, size)
+          }
```

a smaller size should throw an error